File: 2023/python/aoc_2023/day01.py
"""Advent of Code 2023 - Day 1."""

import string
import sys
from typing import TextIO

DIGIT_VALUES = {
    **{d: int(d) for d in string.digits},
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
}


def part_one(file: TextIO) -> int:
    def parse(line: str) -> int:
        digits = [int(c) for c in line if c in string.digits]
        return 10 * digits[0] + digits[-1]

    return sum(map(parse, file))


def part_two(file: TextIO) -> int:
    def parse(line: str) -> int:
        first, first_idx = -1, len(line)
        last, last_idx = -1, -1

        for digit, value in DIGIT_VALUES.items():
            if -1 < line.find(digit) < first_idx:
                first_idx = line.find(digit)
                first = value

            if line.rfind(digit) > last_idx:
                last_idx = line.rfind(digit)
                last = value

        return 10 * first + last

    return sum(map(parse, file))


def main():
    filename = sys.argv[0].replace(".py", ".txt")

    with open(filename, encoding="utf-8") as file:
        print("Part one:", part_one(file))

    with open(filename, encoding="utf-8") as file:
        print("Part two:", part_two(file))


if __name__ == "__main__":
    main()

File: 2023/ruby/day01.rb
# frozen_string_literal: true

input = File.read('input/01.txt').lines(chomp: true)

def part_one(line)
  digits = line.tr('^0-9', '')
  return if digits[0].nil?

  (digits[0] + digits[-1]).to_i
end

puts "Part one: #{input.map(&method(:part_one)).sum}"

$digit_values = {
  "1": 1,
  "one": 1,
  "2": 2,
  "two": 2,
  "3": 3,
  "three": 3,
  "4": 4,
  "four": 4,
  "5": 5,
  "five": 5,
  "6": 6,
  "six": 6,
  "7": 7,
  "seven": 7,
  "8": 8,
  "eight": 8,
  "9": 9,
  "nine": 9
}

def part_two(line)
  first_digit = nil
  first_digit_idx = line.length
  last_digit = nil
  last_digit_idx = -1

  $digit_values.each do |key, value|
    digit = key.to_s
    if line.include?(digit) && line.index(digit) < first_digit_idx
      first_digit_idx = line.index(digit)
      first_digit = value
    end
    if line.include?(digit) && line.rindex(digit) > last_digit_idx
      last_digit_idx = line.rindex(digit)
      last_digit = value
    end
  end

  first_digit * 10 + last_digit
end

puts "Part two: #{input.map(&method(:part_two)).sum}"
